libs/algo/odc/algo/_tiff.py

```diff
@@ -0,0 +1,518 @@
+    settings). Overviews are generated block by block so we do not keep them
+    in-memory uncompressed. To avoid block boundary artefacts, input blocks are
+    set to be 2048x2048 (2**11, edge blocks can have any size). Use that size
+    at construction time for optimal performance.
+
+    :param xx: Geo registered Array, data could be arranged in Y,X or Y,X,B or B,Y,X order.
+               To avoid re-chunking use block sizes of 2048x2048.
+    :param dst: ":mem:" or file path or s3 url
+    :param blocksize: Block size of the final COG (512 pixels)
+    :param ovr_blocksize: Block size of the overview images (default same as main image)
+    :param bigtiff: True|False|"auto" Default is to use bigtiff for inputs greater than 4Gb uncompressed
+    :param temp_folder: By default first pass images are written to RAM, with this option they can be written to disk instead
+    :param overview_resampling: Resampling to use for overview generation: nearest|average|bilinear|...
+
+    :param rio_opts_first_pass: Change defaults for first pass compression
+    :param use_final_blocksizes: By default first pass blocksizes are fixed at
+                                 2048x2048, 1024x1024, 512x512,...64x64, this
+                                 way blocks across different overview levels
+                                 have one to one mapping. With this option one
+                                 can use final image block sizes for the first
+                                 pass instead.
+    :param ACL: Used when dst is S3
+    """
+    assert dask.is_dask_collection(xx)
+    tk = tokenize(
+        xx,
+        dst,
+        blocksize,
+        ovr_blocksize,
+        bigtiff,
+        temp_folder,
+        overview_resampling,
+        extra_rio_opts,
+    )
+
+    info = GeoRasterInfo.from_xarray(xx)
+
+    # Rechunk to 2048x2048 in YX, if needed
+    axis = info.axis
+    data = xx.data
+    chunks = data.chunksize
+    yx_chunks = chunks[axis : axis + 2]
+
+    if yx_chunks != (2048, 2048):
+        data = data.rechunk(chunks[:axis] + (2048, 2048) + chunks[axis + 2 :])
+
+    s3_url: Optional[str] = None
+    extract = False
+    if dst == ":mem:":
+        extract = True
+    elif dst.startswith("s3:"):
+        s3_url, dst = dst, ":mem:"
+    else:
+        # Assume file path
+        # TODO: check if overwrite?
+        # TODO: create folder structure?
+        pass
+
+    # set up sink
+    sink = dask.delayed(COGSink)(
+        info,
+        dst,
+        blocksize=blocksize,
+        ovr_blocksize=ovr_blocksize,
+        bigtiff=bigtiff,
+        temp_folder=temp_folder,
+        overview_resampling=overview_resampling,
+        rio_opts_first_pass=rio_opts_first_pass,
+        use_final_blocksizes=use_final_blocksizes,
+        **extra_rio_opts,
+    )
+
+    rr = da.store(data, sink, lock=False, compute=False)
+
+    # wait for all stores to complete
+    dsk = dict(rr.dask)
+    deps = dsk.pop(rr.key)
```

Comment:
a general dask question, why do you remove `rr.key` from the dask graph here?

Reply:
I'm basically editing graph returned by `da.store` and replacing it's top level result with my own that performs further computation when all chunks were writen

Reply:
I could not just depend on the result of `da.store` because it gets somehow optimized away by Dask, not sure if it's a bug or intended behavior on their part, it was not happening in the older version of Dask, but does now.

Reply:
weird, I think its worth putting in a comment explaining this weirdness